src/uct/ib/rdmacm/rdmacm_cm_ep.c

```diff
@@ -192,27 +289,62 @@ static ucs_status_t uct_rdmacm_cm_create_dummy_qp(struct rdma_cm_id *id,
     return UCS_OK;
 }
 
-static ucs_status_t
-uct_rdamcm_cm_ep_set_qp_num(struct rdma_conn_param *conn_param,
-                            uct_rdmacm_cm_ep_t *cep)
+static ucs_status_t uct_rdamcm_cm_ep_create_qpn(uct_rdmacm_cm_ep_t *cep,
+                                                uint32_t* qpn)
 {
-    struct ibv_cq *cq;
+    struct uct_rdmacm_cm_qp_context *ctx;
     ucs_status_t status;
 
-    status = uct_rdmacm_cm_get_cq(uct_rdmacm_cm_ep_get_cm(cep), cep->id->verbs,
-                                  &cq);
-    if (status != UCS_OK) {
-        return status;
+    status = uct_rdmacm_cm_get_qp_context(uct_rdmacm_cm_ep_get_cm(cep),
+                                          cep->id->verbs, &ctx);
+    ucs_assert(status == UCS_OK);
+
+    if (ctx->reserved_qpn_support) {
+        status = uct_rdamcm_cm_ep_create_reserved_qpn(cep, ctx);
+        if (status == UCS_OK) {
+            cep->flags |= UCT_RDMACM_CM_EP_QPN_CREATED;
+            *qpn = cep->qpn;
+        }        
+    } else {
+        /* create a dummy qp in order to get a unique qp_num to provide to librdmacm */
+        status = uct_rdmacm_cm_create_dummy_qp(cep->id, ctx->dummy_qp_ctx.cq,
+                                               &cep->qp);
+        if (status == UCS_OK) {
+            cep->flags |= UCT_RDMACM_CM_EP_QPN_CREATED;
+            *qpn = cep->qp->qp_num;
+        }
     }
 
-    /* create a dummy qp in order to get a unique qp_num to provide to librdmacm */
-    status = uct_rdmacm_cm_create_dummy_qp(cep->id, cq, &cep->qp);
-    if (status != UCS_OK) {
-        return status;
+    return status;
+}
+
+static void uct_rdamcm_cm_ep_destroy_qpn(uct_rdmacm_cm_ep_t *cep)
+{
+    struct uct_rdmacm_cm_qp_context *ctx;
+    ucs_status_t status;
+
+    if (!(cep->flags & UCT_RDMACM_CM_EP_QPN_CREATED)) {
+        return;
     }
 
-    conn_param->qp_num = cep->qp->qp_num;
-    return UCS_OK;
+    status = uct_rdmacm_cm_get_qp_context(uct_rdmacm_cm_ep_get_cm(cep),
+                                          cep->id->verbs, &ctx);
+    ucs_assert_always(status == UCS_OK);
```

Comment:
can we propagate the error code to a caller API? `ucs_assert_always` means fatal error, probably this is not the case

Reply:
@evgeny-leksikov Not good to do it in the two functions calling this function. One is a cleanup function. The other calls this function in error handling which it deals the status error of other functions.
In destroy function get qp context failure is unexpected since the context should already in hash table, maybe we can treate it as fatal error.

Reply:
need to avoid fatal errors where it's not actually needed, pls handle error gracefully
if we can not return error to user, print error

Reply:
@evgeny-leksikov Fixed